File: Score Counter Animated - Example Project/Score Counter Animated/Graph/ScriptTemplate/Counter/CGCounter.js
const {BaseNode} = require('./BaseNode');
const Amaz = effect.Amaz;

class CGCounter extends BaseNode {
  constructor() {
    super();
    this.value = 0;
  }

  beforeStart(sys) {
    this.value = this.inputs[4]();
  }

  inRange(from, to, value) {
    if (from <= to) {
      return from <= value && value <= to;
    } else {
      return from >= value && value >= to;
    }
  }

  clamp(from, to, value) {
    if (from <= to) {
      return Math.min(to, Math.max(from, value));
    } else {
      return Math.min(from, Math.max(to, value));
    }
  }

  execute(index) {
    const step = this.inputs[3]();
    const from = this.inputs[4]();
    const to = this.inputs[5]();

    this.value = this.clamp(from, to, this.value);

    if (index === 0) {
      if (this.inRange(from, to, this.value + step)) {
        this.value = this.value + step;
      }
    } else if (index === 1) {
      if (this.inRange(from, to, this.value - step)) {
        this.value = this.value - step;
      }
    } else if (index === 2) {
      this.value = from;
    }
    if (this.nexts[0]) {
      this.nexts[0]();
    }
  }

  getOutput(index) {
    return this.value;
  }
}

exports.CGCounter = CGCounter;

File: Score Counter Animated - Example Project/Score Counter Animated/Graph/ScriptTemplate/Counter/CGCounter.lua
local CGCounter = CGCounter or {}
CGCounter.__index = CGCounter

function CGCounter.new()
    local self = setmetatable({}, CGCounter)
    self.inputs = {}
    self.outputs = {}
    self.nexts = {}
    self.value = nil
    return self
end

function CGCounter:setNext(index, func)
    self.nexts[index] = func
end

function CGCounter:setInput(index, func)
    self.inputs[index] = func
end

function CGCounter:getOutput(index)
    if self.outputs[index] then
        return self.outputs[index]
    end
    if self.inputs[2]() < self.inputs[3]() or self.inputs[2]() > self.inputs[4]() then
        return nil
    end
    return self.inputs[2]()
end

function CGCounter:execute(index)
    if self.inputs[2] == nil then
        return
    end
    -- stuck here if base value out of range
    if self.inputs[2]() < self.inputs[3]() or self.inputs[2]() > self.inputs[4]() then
        return
    end

    if index == 0 then
        if self.value == nil then
            self.value = self.inputs[2]()
        end
        self.value = self.value + 1
    end

    if index == 1 then
        if self.value == nil then
            self.value = self.inputs[2]()
        end
        self.value = self.value - 1
    end

    if self.value > self.inputs[4]() then
        self.value = self.inputs[3]()
    end
    if self.value < self.inputs[3]() then
        self.value = self.inputs[4]()
    end
     
    self.outputs[1] = self.value

    if self.nexts[0] then
        self.nexts[0]()
    end
end

return CGCounter
